### PR TITLE
vim-patch:9.0.{0175,0590,0608,0664}

### LINKED ISSUE #11285: Incorrect redraw in editing a long line with C when 'spell' is enabled
### Steps to reproduce

1. Enable spell check (`set spell`)
2. Edit a file containing a line longer than window size, followed by a series of lines with at least one blank line (see attached a.txt)
3. Go to the middle of that long line and type C to enter insert mode.
4. Observe that subsequent lines disappear with each keystroke.

See attached a.mp4 for a demo.

This happens in both console vim and gvim.

### Expected behaviour

Entering insert mode and replacing the suffix of the current line should not affect subsequent lines.

Disabling spell check (`set nospell`) and typing C results in a $ to be drawn at the end of the current line, and editing the line suffix does not affect subsequent lines.  This is the expected behavior.

### Version of Vim

9.0.662

### Environment

OS = Linux version 5.18.16-1rodete1-amd64 (glinux-team@google.com) (gcc-11 (Debian 11.3.0-5) 11.3.0, GNU ld (GNU Binutils for Debian) 2.38.90.20220713) #1 SMP PREEMPT_DYNAMIC Debian 5.18.16-1rodete1 (2022-08-24)
Terminal = # GNOME Terminal 3.44.1 using VTE 0.69.99 +BIDI +GNUTLS +ICU +SYSTEMD
TERM = xterm-256color
Shell = tcsh

https://user-images.githubusercontent.com/2000051/193947690-fdf95d06-ed6e-47f8-a648-6d6ccf21594e.mp4

[a.txt](https://github.com/vim/vim/files/9711422/a.txt)

### Logs and stack traces

_No response_

### PR DESCRIPTION
**Pull request by luukvbaal**
**vim-patch:9.0.0175: spell checking for capital not working with trailing space**

Problem:    Spell checking for capital not working with trailing space.
Solution:   Do not calculate cap_col at the end of the line. (Christian
            Brabandt, closes vim/vim#10870, issue vim/vim#10838)

https://github.com/vim/vim/commit/afa23d1b99692e3c726eb694933ab348b442a1e4

**vim-patch:9.0.0590: after exiting Insert mode spelling not checked in next line**

Problem:    After exiting Insert mode spelling is not checked in the next
            line.
Solution:   When spelling is enabled redraw the next line after exiting Insert
            mode in case the spell highlight needs updating.

https://github.com/vim/vim/commit/ee09fcc9b6cf24e02899461809da9a5148208ea5

**vim-patch:9.0.0608: with spelling, deleting a full stop does not update next line**

Problem:    With spell checking, deleting a full stop at the end of a line
            does not update SpellCap at the start of the next line.
Solution:   Update the next line when characters have been deleted.  Also when
            using undo.

https://github.com/vim/vim/commit/26f09ea54b2c60abf21df42c60bdfc60eca17b0d

**vim-patch:9.0.0664: bad redrawing with spell checking, using "C" and "$" in 'cpo'**

Problem:    Bad redrawing with spell checking, using "C" and "$" in 'cpo'.
Solution:   Do not redraw the next line when "$" is in 'cpo'. (closes vim/vim#11285)

https://github.com/vim/vim/commit/f3ef026c9897f1d2e3fba47166a4771d507dae91

### CODE REVIEW

**Comment by luukvbaal** [comment]
Batch of vim patches related to https://github.com/vim/vim/pull/12428. None of the code changes here are actually required to port that but I only later realized I modified an as of yet unported test there so I though I might as well port these...

vim-patch:9.0.0048 also seems largely N/A because the original problem does not exist in nvim. We already properly account for the offset with our screen grid. Hence, I only ported the test from that patch, I'll check one more time if any of the code/comment changes are relevant.

**Comment by zeertzjq** [comment]
> We already properly account for the offset with our screen grid.

I don't think screen grid does things related to buffer content.

**Comment by luukvbaal** [comment]
I didn't look into what's actually responsible for adjusting the cursor position so yeah that statement might be incorrect. The point is that the cursor position is correctly adjusted when clicking beyond concealed text.

**Comment by zeertzjq** [comment]
Unlike the approach in the Vim patch, `mouse_adjust_col()` may need to be adjusted if the redrawing code changes.

**Comment by zeertzjq** [comment]
Patch 9.0.0048 doesn't seem required by any following patches in this PR, so I think it's better to drop that patch from this PR.

**Comment by luukvbaal** [comment]
True, the code changes in that patch were loosely related with the introduction of `prev_ptr` but it didn't turn out to be necessary for the other patches. Since I didn't end up porting the code change should I still open a different PR with just the ported test or just leave it be?

**Comment by zeertzjq** [comment]
The test in Vim script already work with `Ntest_setmouse()`, and `mouse_adjust_click()` is already covered by tests `ui/mouse_spec.lua`, so it isn't really urgent to port the test when nothing else depends on it.

**Comment by luukvbaal** [comment]
> The test in Vim script already work with Ntest_setmouse(), and mouse_adjust_click() is already covered by tests ui/mouse_spec.lua, so it isn't really urgent to port the test when nothing else depends on it.

Right forgot about that and assumed without checking that `Ntest_setmouse()` would just return early for some reason. And yeah, makes sense.

File: src/nvim/spell.c [hunk 1]
```diff
@@ -1189,11 +1189,19 @@ bool spell_valid_case(int wordflags, int treeflags)
                  || (wordflags & WF_ONECAP) != 0));
 }
 
-// Returns true if spell checking is not enabled.
+/// Return true if spell checking is enabled for "wp".
+int spell_check_window(win_T *wp)
```
Comment (zeertzjq) [inline src/nvim/spell.c:1193]:
```suggestion
bool spell_check_window(win_T *wp)
```

**Review by zeertzjq — APPROVED** [review]

**Comment by luukvbaal** [comment]
Do we backport this?

**Comment by luukvbaal** [comment]
I guess vim-patch:9.0.0218 from #23612 would first need to be backported make it apply cleanly. If that doesn't apply cleanly either I guess it's whatever.

**Pull request merged**